### PR TITLE
PS-3375 Add `ManageClient::listProjects`

### PR DESCRIPTION
**Pull request by joker806**
https://keboola.atlassian.net/browse/PS-3375
Sandboxes API PR: https://github.com/keboola/sandboxes-api/pull/179

### CODE REVIEW

File: src/ManageClient.php [hunk 1]
```diff
@@ -62,6 +62,26 @@ public function deactivate(string $id): void
...
+        do {
+            $query['nextPageToken'] = $nextPageToken;
+            $response = $this->sendRequest(new Request('GET', '/manage/projects?'.http_build_query($query)));
+
+            foreach ($response['data'] ?? [] as $projectData) {
+                yield Project::fromArray($projectData);
+            }
+
+            $nextPageToken = $response['nextPageToken'] ?? '';
+        } while (!empty($nextPageToken));
```
Comment (pivnicek) [inline src/ManageClient.php:82]: oooh, an old school do - while loop :)

File: tests/ListProjectsOptionsTest.php [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+
+        yield 'hasPersistentStorage: true' => [
+            ListProjectsOptions::create()->setHasPersistentStorage(true),
+            ['hasPersistentStorage' => true],
+        ];
+
+        yield 'hasPersistentStorage: false' => [
+            ListProjectsOptions::create()->setHasPersistentStorage(false),
+            ['hasPersistentStorage' => false],
+        ];
```
Comment (pivnicek) [inline tests/ListProjectsOptionsTest.php:35]: how about adding here one case for where `clearHasPersistentStorage` has been called?

File: tests/ManageClientTest.php [hunk 3]
```diff
@@ -53,4 +55,72 @@ function (RequestInterface $request): ResponseInterface {
...
+                return new Response(200, [], (string) json_encode([
+                    'data' => [
+                        [
+                            'id' => '111',
+                            'projectId' => 'project1',
+                            'tokenId' => 'token-id',
+                            'type' => 'python',
+                            'active' => true,
+                            'createdTimestamp' => '2021-12-12T12:00:00Z',
+                        ],
```
Comment (pivnicek) [inline tests/ManageClientTest.php:77]: if these have persistent storage, then I guess they should return it?  I know it's just a mock, ale ... 😄

**Review by pivnicek — APPROVED** [review]
LGTM, just a couple of qs in the tests

File: tests/ListProjectsOptionsTest.php (hunk 2 above)
Comment (joker806) [reply tests/ListProjectsOptionsTest.php:35]: I don't think it belongs here. The `testToQueryParameters` tests if various states of the `ListProjectsOptions` are properly exported. And all possible state are already covered (`hasPersistentStorage: null/true/false`).

`set/get/clear` behaviour is tested by `testHasPersistentStorage`.

File: tests/ManageClientTest.php (hunk 3 above)
Comment (joker806) [reply tests/ManageClientTest.php:77]: I don't really have any reasons here :slightly_smiling_face: On one hand, you're righ it does not represent how project looks in reality. But on the other hand, this shows the minimal needed configuration for the code to work and adding the persistent storage to data may lead you to a wrong assumption that the data actually matters.

Comment (joker806) [reply tests/ManageClientTest.php:77]: Ok, after thinking of it for 30 more second, I'll add it :smile:

File: src/ManageClient.php (hunk 1 above)
Comment (joker806) [reply src/ManageClient.php:82]: It's not that old school. I use this patter quite often when hiding API pagination using generators and I don't consider myself oldschool :wink:

File: tests/ManageClientTest.php (hunk 3 above)
Comment (joker806) [reply tests/ManageClientTest.php]: Done

**Pull request merged**